elixir/apps/domain/lib/domain/events/hooks/accounts.ex

```
@@ -1,13 +1,42 @@
 defmodule Domain.Events.Hooks.Accounts do
+  alias Domain.PubSub
+  require Logger
+
   def on_insert(_data) do
     :ok
   end
 
-  def on_update(_old_data, _data) do
-    :ok
+  def on_update(old_data, data) do
+    with {:ok, account_id} <- Map.fetch(data, "id"),
+         {:ok, old_config} <- Map.fetch(old_data, "config"),
+         {:ok, config} <- Map.fetch(data, "config") do
+      if old_config != config do
+        broadcast(account_id, :config_changed)
+      else
+        :ok
+      end
+    else
+      :error ->
+        Logger.error("Expected keys not found in data",
+          old_data: inspect(old_data),
+          data: inspect(data)
+        )
+
+        :ok
+    end
   end
```

I noticed there's a decent amount control flow in this function.  Would it be more of "the Elixir way" to utilize pattern matching here?  Something like:

```
def on_update(%{"config" => old_config} , %{"id" => account_id, "config" => config}) do
	if old_config != config do
		broadcast(account_id, :config_changed)
	else
		:ok
	end
end

def on_update(old_data, data) do
	Logger.error("Expected keys not found in data",
          old_data: inspect(old_data),
          data: inspect(data)
        )
	:ok
end
```

---

I thought about that, but I'm not sure. These handlers could get quite complicated if we try to handle this in the function head. I thought it might be more maintainable / easier to follow if we keep a single function per event handler and then handle the side effects in the function body.

I foresee us adding helpers to parse things and check for equality once a pattern emerges, for example.

Another thing I though might be a good idea is to instantiate structs out of these so that we get some level of type checking for the missing fields. Wdyt?

---

Yeah, I like the idea of structs to allow for some level of type checking.

I can understand your concern about pattern matching in the function args if there are going to be a bunch of variations we're going to have to create.  If we can create structs, it seems like that might be able to help get some of the conditional checking out of these functions and let them deal with just the effects logic they need to deal with and worry about the verifying logic in another way.

---

Ah, structs can't be instantiated from string keys directly. I suppose matching on needed fields is probably the best approach then. I think we can let it crash if the other fields aren't there as this should be quite rare and that is an alarm we want to be noisy about.

---

The right approach will make itself clear after I do a few more of these.

---

Updated